### PR TITLE
4750 contacts report component

### PR DESCRIPTION
**Pull request by njogz**
# Description

This migrates the contacts report component

medic/cht-core#[number]

# Code review items

- Readable: Concise, well named, follows the [style guide](https://docs.communityhealthtoolkit.org/contribute/code/style-guide/), documented if necessary.
- Documented: Configuration and user documentation on [cht-docs](https://github.com/medic/cht-docs/)
- Tested: Unit and/or e2e where appropriate
- Internationalised: All user facing text
- Backwards compatible: Works with existing data and configuration or includes a migration. Any breaking changes documented in the release notes.

# License

The software is provided under AGPL-3.0. Contributions to this project are accepted under the same license.

### CODE REVIEW

File: webapp/src/ts/modules/contacts/contacts-report.component.ts [hunk 1]
```diff
@@ -0,0 +1,243 @@
...
+        return;
+      }
+      this.routeSnapshot = this.route.snapshot;
+      this.resetState();
+      this.resetFormError();
+      this.render();
+    });
+    this.subscription.add(routeSubscription);
+
+    const queryParamsSubscription = this.route.queryParams.subscribe(() => {
```
Comment (dianabarsan) [inline webapp/src/ts/modules/contacts/contacts-report.component.ts:183]: Is it possible to get queryParams here? Which query params are you expecting?

**Review by dianabarsan — CHANGES_REQUESTED** [review]
Great work! 

I have some comments inline. 

Also, I've noticed that you didn't implement the canDeactivate route guard, that shows the "cancel" modal when navigating away from an edited form. Can you, please implement that?

Also, can you please rebase off of the main branch and incorporate the necessary updates of the actionbar in this PR? it would make my testing much easier! 

Thanks!!

File: webapp/src/ts/modules/contacts/contacts-report.component.ts [hunk 2]
```diff
@@ -0,0 +1,243 @@
...
+  ngOnInit() {
+    this.subscribeToStore();
+    this.subscribeToRoute();
+
+    this.geoHandle = this.geolocationService.init();
+    this.resetFormError();
+    this.form = null;
+    this.loadingForm = true;
+    this.globalActions.setShowContent(true);
+    this.setCancelCallback();
```
[the remark this answers is not in the document]
Comment (njogz) [reply webapp/src/ts/modules/contacts/contacts-report.component.ts:64]: Done.

File: webapp/src/ts/modules/contacts/contacts-report.component.ts [hunk 3]
```diff
@@ -0,0 +1,243 @@
...
+    const queryParamsSubscription = this.route.queryParams.subscribe(() => {
+      this.routeSnapshot = this.route.snapshot;
+      this.setCancelCallback();
+    });
+    this.subscription.add(queryParamsSubscription);
+  }
+
+  private setCancelCallback() {
+    this.routeSnapshot = this.route.snapshot;
+    if (this.routeSnapshot.params && (this.routeSnapshot.params.id || this.routeSnapshot.params.formId)) {
```
[the remark this answers is not in the document]
Comment (njogz) [reply webapp/src/ts/modules/contacts/contacts-report.component.ts]: You are right `formId` does not make much sense here. I have made the suggested change.

File: webapp/src/ts/modules/contacts/contacts-report.component.ts [hunk 4]
```diff
@@ -0,0 +1,243 @@
...
+    }
+
+    this.telemetryData.preSave = Date.now();
+    this.telemetryService.record(
+      `enketo:contacts:${this.telemetryData.form}:add:user_edit_time`,
+      this.telemetryData.preSave - this.telemetryData.postRender);
+
+    this.globalActions.setEnketoSavingStatus(true);
+    this.resetFormError();
+    this.enketoService.save(this.routeSnapshot.params.formId, this.form, this.geoHandle)
```
[the remark this answers is not in the document]
Comment (njogz) [reply webapp/src/ts/modules/contacts/contacts-report.component.ts]: Done.

File: webapp/src/ts/modules/contacts/contacts-report.component.ts [hunk 5]
```diff
@@ -0,0 +1,243 @@
...
+      .then((docs) => {
+        console.debug('saved report and associated docs', docs);
+        this.globalActions.setEnketoSavingStatus(false);
+        this.globalActions.setSnackbarContent(this.translateService.instant('report.created'));
+        this.globalActions.setEnketoEditedStatus(false);
+        this.router.navigate([`/contacts/${this.routeSnapshot.params.id}`]);
```
[the remark this answers is not in the document]
Comment (njogz) [reply webapp/src/ts/modules/contacts/contacts-report.component.ts]: Done.

File: webapp/src/ts/modules/contacts/contacts-report.component.ts [hunk 6]
```diff
@@ -0,0 +1,243 @@
...
+        this.telemetryData.postSave = Date.now();
+
+        this.telemetryService.record(
+          `enketo:contacts:${this.telemetryData.form}:add:save`,
+          this.telemetryData.postSave - this.telemetryData.preSave);
+      })
+      .catch((err) => {
+        this.globalActions.setEnketoSavingStatus(false);
+        console.error('Error submitting form data: ', err);
+        this.globalActions.setEnketoError(this.translateService.get('error.report.save'));
```
[the remark this answers is not in the document]
Comment (njogz) [reply webapp/src/ts/modules/contacts/contacts-report.component.ts]: Meant `.instant` in deed. Changed.

File: webapp/src/ts/modules/contacts/contacts-report.component.ts [hunk 7]
```diff
@@ -0,0 +1,243 @@
...
+    this.subscription.unsubscribe();
+    this.geoHandle && this.geoHandle.cancel();
+    this.enketoService.unload(this.form);
+  }
+
+  ngAfterViewInit() {
+    this.render();
+  }
+
+  private getContact() {
```
[the remark this answers is not in the document]
Comment (njogz) [reply webapp/src/ts/modules/contacts/contacts-report.component.ts:78]: Yes. I have refactored the code to use `ContactViewModelGenerator`.

File: webapp/src/ts/modules/contacts/contacts-report.component.ts (hunk 1 above)
Comment (njogz) [reply webapp/src/ts/modules/contacts/contacts-report.component.ts]: Not necessary. I have removed the code block.

**Comment by njogz** [comment]
@dianabarsan thanks for the feedback. I have addressed your comments. Please have another look when you can.

**Review by dianabarsan — CHANGES_REQUESTED** [review]
I'm noticing that the correct contact is not getting associated with the submitted report. 
To reproduce: 
- load a contact
- open a new form that uses this contact (for example, a delivery or pregnancy form)
- fill the form and submit
- notice that the report doesn't appear on the contact's detail page.
- navigate to reports and see that the report has empty `patient_id` and `patient_uuid` fields.

File: webapp/tests/karma/ts/modules/contacts/contacts-report.component.spec.ts [hunk 8]
```diff
@@ -143,6 +143,24 @@ describe('contacts report component', () => {
...
+
+      expect(enketoService.render.callCount).to.equal(1);
+      console.log(enketoService.render.args[0][3]);
+      expect(enketoService.render.args[0][0]).to.equal('#contact-report');
+      expect(enketoService.render.args[0][1]).to.deep.equal({ title: 'formTitle' });
+      expect(enketoService.render.args[0][2]).to.deep.equal(
+        {
+          source: 'contact',
+          contact: {
+            doc: {}
```
Comment (dianabarsan) [inline webapp/tests/karma/ts/modules/contacts/contacts-report.component.spec.ts:158]: Can we please enrich this test with some mock data?
At this time, I have absolutely no idea where the empty doc object is coming from.

**Review by dianabarsan — CHANGES_REQUESTED** [review]
@njogz I'm seeing this error when I try to submit an "invalid" form twice. 

I'm using a home_visit form and I'm deliberately not filling in the home visit date and clicking submit. 1st click does what I'd expect it to do: it reports that the form is invalid and changes the markup, but on second click this happens: 

image

File: webapp/src/ts/modules/contacts/contacts-report.component.ts [hunk 9]
```diff
@@ -0,0 +1,241 @@
...
+    }
+  }
+
+  private markFormEdited() {
+    this.globalActions.setEnketoEditedStatus(true);
+  }
+
+  private resetFormError() {
+    if (this.enketoError) {
+      this.globalActions.setEnketoError.setEnketoError(null);
```
Comment (dianabarsan) [inline webapp/src/ts/modules/contacts/contacts-report.component.ts:103]: This line is causing the error described above. 
Please add a unit test for this.

File: webapp/tests/karma/ts/modules/contacts/contacts-report.component.spec.ts (hunk 8 above)
Comment (njogz) [reply webapp/tests/karma/ts/modules/contacts/contacts-report.component.spec.ts]: Done.

File: webapp/src/ts/modules/contacts/contacts-report.component.ts (hunk 9 above)
Comment (njogz) [reply webapp/src/ts/modules/contacts/contacts-report.component.ts]: Good catch. Fixed.

**Review by dianabarsan — APPROVED** [review]
Nice work!

**Pull request merged**